timemachine/fe/dummy.py

```diff
@@ -1,151 +1,202 @@
 
-    cc = list(nx.connected_components(induced_g))
+    Parameters
+    ----------
+    mol_a, mol_b: Mol
+        Source and target molecules for alchemical transformation.
+        Dummy atoms are added to mol_a to transform it into a supergraph of mol_b.
 
-    # check that we didn't miss any dummy atoms
-    assert np.sum([len(c) for c in cc]) == (N - len(core))
+    core_atoms_a: collection of int
+        mol_a atoms in the core
 
-    return cc
+    core_atoms_b: collection of int
+        mol_b atoms in the core
 
+    Returns
+    -------
+    iterator of dict
+        each element is a mapping from bond anchor atom to the pair (angle anchor atom, dummy group)
+    """
 
-def canonicalize_bond(ixn):
+    bonds_b = get_romol_bonds(mol_b)
+    bond_graph_b = convert_bond_list_to_nx(bonds_b)
+    dummy_group_assignments = generate_dummy_group_assignments(bond_graph_b, core_atoms_b)
+
+    core_bonds_c = get_core_bonds(mol_a, mol_b, core_atoms_a, core_atoms_b)
+    c_to_b = {c: b for c, b in enumerate(core_atoms_b)}
+    core_bonds_b = frozenset(translate_bonds(core_bonds_c, c_to_b))
+
+    def get_angle_anchors(bond_anchor):
+        valid_angle_anchors = [
+            angle_anchor
+            for angle_anchor in [n for n in bond_graph_b.neighbors(bond_anchor) if n in core_atoms_b]
+            if canonicalize_bond((bond_anchor, angle_anchor)) in core_bonds_b
+        ]
+        return valid_angle_anchors or [None]
+
+    anchored_dummy_group_assignments = (
+        dict(anchored_dummy_group)
+        for dummy_group_assignment in dummy_group_assignments
+        for anchored_dummy_group in product(
+            *[
+                [(bond_anchor, (angle_anchor, dummy_group)) for angle_anchor in get_angle_anchors(bond_anchor)]
+                for bond_anchor, dummy_group in dummy_group_assignment.items()
+            ]
+        )
+    )
+
+    return anchored_dummy_group_assignments
+
+
+def canonicalize_bond(ixn: Tuple[int, ...]) -> Tuple[int, ...]:
     if ixn[0] > ixn[-1]:
         return tuple(ixn[::-1])
     else:
         return tuple(ixn)
+
+
+def convert_bond_list_to_nx(bond_list: Collection[Tuple[int, int]]) -> nx.Graph:
+    """
+    Convert an ROMol into a networkx graph.
+    """
+    g = nx.Graph()
+    for i, j in bond_list:
+        assert i != j  # do not allow self-edges
+        g.add_edge(i, j)
+
+    return g
+
+
+def get_core_bonds(
+    mol_a, mol_b, core_atoms_a: Collection[int], core_atoms_b: Collection[int]
+) -> FrozenSet[Tuple[int, int]]:
+    """Returns core-core bonds that are present in both mol_a and mol_b"""
+    bonds_a = get_romol_bonds(mol_a)
+    bonds_b = get_romol_bonds(mol_b)
+    a_to_c = {a: c for c, a in enumerate(core_atoms_a)}
+    b_to_c = {b: c for c, b in enumerate(core_atoms_b)}
+    return frozenset(translate_bonds(bonds_a, a_to_c)).intersection(frozenset(translate_bonds(bonds_b, b_to_c)))
+
+
+def translate_bonds(bonds: Collection[Tuple[int, int]], mapping: Dict[int, int]):
```

Comment:
nit: docstring

Reply:
Added in https://github.com/proteneer/timemachine/pull/1019/commits/285bdc753592abbe038415f7a3041d39cc7c7feb